### PR TITLE
Migrate from python-ldap to ldap3

### PR DESCRIPTION
**Pull request by mprahl**
* Turn off logging during tests
* Replace python-ldap code with ldap3
* Simplifies the check_nested_group_membership function so that the membership checking is done in AD and not in Python
* Since ldap3 allows NTLM authentication against AD, this was added as a config option
* Unit tests greatly simplified

This is a big PR, I think it'd be easier for you to review on a per commit basis (I tried to make those neat and concise changes).

### CODE REVIEW

File: postmaster/ad.py [hunk 1]
```diff
@@ -0,0 +1,229 @@
...
+        """ Takes a username and properly formats it for authentication with Active Directory
+        """
+        extract_username_regex = re.compile(r'(?P<username>.+)(?:@.*)')
+        # Determine if a UPN or domain was provided
+        if '\\' in username:
+            return username
+        elif re.match(extract_username_regex, username):
+            # Parse the username from the UPN
+            username_search = re.search(extract_username_regex, username)
+            return self.domain + '\\' + username_search.group('username')
```
Comment (thatarchguy) [inline postmaster/ad.py]: Either use `.format` or `.join` to format strings.

File: postmaster/ad.py [hunk 2]
```diff
@@ -0,0 +1,229 @@
...
+        elif self.ldap_connection.authentication != 'NTLM' and re.search('CN=', username, re.IGNORECASE):
+            # If the authentication method is not NTLM, then distinguished names are valid usernames
+            return username
+        else:
+            return self.domain + '\\' + username
```
Comment (thatarchguy) [inline postmaster/ad.py]: Either use `.format` or `.join` to format strings.

File: postmaster/utils.py [hunk 3]
```diff
@@ -254,6 +256,8 @@ def add_ldap_user_to_db(username, display_name):
 def validate_wtforms_password(form, field):
         """ Validates the password from a wtforms object
         """
+        # Prevent circular import on json_logger by importing here
+        from postmaster.ad import AD, ADException
```
Comment (thatarchguy) [inline postmaster/utils.py]: How does this create a circular import on json_logger?

File: tests/ad/ad_ldap_directory.json [hunk 4]
```diff
@@ -0,0 +1,912 @@
+{
+    "entries": [
```
Comment (thatarchguy) [inline tests/ad/ad_ldap_directory.json]: My oh my I hope this was exported from somewhere!

**Comment by thatarchguy** [comment]
Beautiful merge request. I can only really nitpick at one or two small things. The tests are good and the code is good. I really appreciate the hard work you put into that merge. 

I think the way we are handling config stuff in the database is getting a bit big. Controllable though, just something we may need to keep an eye on in the future.

File: tests/ad/ad_ldap_directory.json (hunk 4 above)
Comment (mprahl) [inline tests/ad/ad_ldap_directory.json]: Fortunately, yes :)

Here's an example from the [docs](http://ldap3.readthedocs.io/mocking.html):
```
from ldap3 import Server, Connection, ALL_ATTRIBUTES
server = Server('my_real_server')
connection = Connection(server, 'cn=my_real_user,ou=test,o=lab', 'my_real_password', auto_bind=True)
if connection.search('ou=test,o=lab', '(objectclass=*)', attributes=ALL_ATTRIBUTES):
    connection.response_to_file('my_entries.json', raw=True)
```

File: postmaster/ad.py (hunk 1 above)
Comment (mprahl) [inline postmaster/ad.py]: Good point, I'll fix this tonight.

File: postmaster/ad.py (hunk 2 above)
Comment (mprahl) [inline postmaster/ad.py]: Good point, I'll fix this tonight.

File: postmaster/utils.py (hunk 3 above)
Comment (thatarchguy) [inline postmaster/utils.py]: We may run into this in the future. It may be better to separate the logger out into its own logger.py file.

**Comment by mprahl** [comment]
@thatarchguy thoughts on the last commit?

**Comment by thatarchguy** [comment]
[LGTM](https://lgtm.in/i/EqHN1uGPp)

[:+1:](https://lgtm.in/u/EqHN1uGPp)

**Pull request merged**